[WIP DO NOT MERGE] Migrate ZCL over message layer

kghost opened the pull request:
Open for discussion, do not merge

 #### Problem
As we already discussed that use message layer to dispatch packets to ZCL cluster.

 #### Summary of Changes
Modify chip-tool and examples to use message layer API

andy31415 commented:
@kghost I converted this to a draft based on the title. I hope that is ok. If not, please undo and update the title.

bzbarsky-apple commented on src/app/util/af-main-common.cpp:
```diff
@@ -227,7 +227,7 @@ EmberAfCbkeKeyEstablishmentSuite emberAfIsFullSmartEnergySecurityPresent(void)
 }
 
-static EmberStatus send(EmberOutgoingMessageType type, uint64_t indexOrDestination, EmberApsFrame * apsFrame,
+static EmberStatus send(EmberOutgoingMessageType type, void * exchangeContext, EmberApsFrame * apsFrame,
```
This does not seem right to me.  It'll work OK for responses, but not for server-initiated communication (like attribute reports, events, etc, etc).

Now maybe such communication needs to set up an exchange somewhere higher up the stack, and maybe we need to rejigger the APIs to more clearly differentiate between responses and server-initiated communication....

It looks like this PR just rips out a bunch of existing functionality to do with bindings and whatnot.  We should instead design things so that this functionality works naturally; there is no obvious differences between outgoing communication to a specific node id and to a binding, except insofar as whether we're doing a response or not...

Please also note https://github.com/project-chip/connectedhomeip/issues/3584 -- it might be worth addressing that and then having that type be able to indicate whether this send is part of an existing exchange or the start of a new exchange.

kghost replied:
I'd like to move the binding concept to exchange layer, this is why I purged all binding related code in ember.

Ideally, messages are exchanged using an `ExchangeContext`, for both clients and servers (controllers and devices). The binding or peer node id will be hidden inside the `ExchangeContext`.

The client need to create the context manually, using given binding (either discovered by mDNS or specified via command line). Once the context is created, upper layer can use the context directly w/o knowing anything about binding or addressing.

The server is more simple. Once it receive a message from client requesting an exchange, it will automatically allocate and set up an `ExchangeContext`, and pass it to applications or protocols. The upper layer can use it directly w/o knowing anything about binding or addressing.

I'll highlight some essential part of the  PR to help understand the flow.

bzbarsky-apple replied:
> I'd like to move the binding concept to exchange layer

You can't do that; it's a data model concept....

> The binding or peer node id will be hidden inside the ExchangeContext

And exchange context is transient, with the lifetime of an exchange, right?  A binding is long-lived; it persists across power cycles, etc.

> The client need to create the context manually, using given binding (either discovered by mDNS or specified via command line)

Maybe we are talking about different meanings of "binding"?

> The server is more simple

The server can also have bindings, no?

kghost replied:
Thanks for the comment. IMO, binding translates application discriminator, attribute or configuration of a connection into under layer information.

Bindings are constructed by applications, and may contains following attributes:
 * peer node id or group id for multicast
 * type: unicast or multicast
 * authentication method: SPAKE or SIGMA or No encryption
 * required or preferred transports: BLE, WiFi or Thread

These kind of information will help all under layer components select proper configurations to set up contexts. It decoupled internal implementation details of under layer components from application. For example, application doesn't need to know the address of a peer node.

Also, I'd like to let the binding focus on initializing the message channel. All information inside bindings will be resolved to actual data, and stored in contexts, application don't need to use the binding after the establishment. an `ExchangeContext` will be provide to application, use it to exchange following messages.

For the server side, we can derive the context from the initiator message, so binding may not be mandatory. And if application want to persist the information of peer nodes, there will be APIs to retrieve data from contexts to reconstruct the binding representing peer node, like `GetPeerNodeId`, `IsExchangeSecure`, `GetTransportType` and so on.

In general. Use binding to initiate a exchange channel, use context to send/receive messages.

bzbarsky-apple replied:
> IMO, binding translates application discriminator, attribute or configuration of a connection into under layer information

"binding" has a specific definition in a CHIP context.  It translates an (endpoint id, cluster id) pair into a list of (group id or node id, endpoint id) pairs.  That is, it's a list of endpoints that are bound to the given endpoint for the given cluster.

https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/cam/system-model-spec/src/data_model/System-Model.adoc#3-binding-relationship and https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/cam/system-clusters/src/data_model/Binding-Cluster.adoc for some more information, though they are not very polished so far....

> and may contains following attributes

From the point of view of the data model, the "authentication method" is a lower-level implementation detail at the moment, and not captured in the data model concept of a "binding".  If it needs to be persistently stored at the data model level, we should adjust the data model bits accordingly....

Same for the transport bits.

> For example, application doesn't need to know the address of a peer node.

If by "address" you mean IP address, absolutely.  As you might note from the PRs above, the data model uses "address" to mean "Node id" in various cases.  It should probably stop doing that... 

> application don't need to use the binding after the establishment

Yes, that makes sense.

> For the server side, we can derive the context from the initiator message

The "server" can be the initiator, is the point.  At least if you mean the server side of a cluster.  Again, we might be using different meanings of the term "server" here.  :)

> In general. Use binding to initiate a exchange channel, use context to send/receive messages.

This makes perfect sense.

Again, in the low-level data model sending code in `af-main-common.cpp` it feels to me like we should support "send a message here" where "here" might be a binding or an existing exchange.  Whether we need support for "here" being a raw node id is an interesting question.  In past discussion, it seemed that a node id could be used to look up the relevant exchange context, which is why the code is structured the way it is right now.  But I would be fine with explicitly passing through some sort of exchange identifier when we know we are processing a message from an exchange (and hence responses should go in the same exchange) instead; whatever makes sense to the lower levels here.

So the way I see it, when we are done `af-main-common` message-sending code examines its arguments.  If it's provided an existing exchange, it uses that to send the message.  If it's provided a binding, it initiates a new exchange and sends on that.  Whether it also needs explicit nodeid/groupid forms (as it has now) or whether the "exchange" form would subsume those is TBD; we need to audit the uses of those forms to see what situations they are meant to cover.

Does that all make sense?

kghost commented on examples/chip-tool/commands/common/EchoCommand.cpp:
```diff
@@ -78,20 +80,35 @@ void EchoCommand::ReceiveEcho(PacketBuffer * buffer) const
+    em = dc->GetExchangeManager();
+    VerifyOrExit(em != nullptr, err = CHIP_ERROR_NO_MEMORY);
+
+    ec = em->NewContext(remoteId, this);
+    VerifyOrExit(ec != nullptr, err = CHIP_ERROR_NO_MEMORY);
```
Client (AKA controller) needs to manually create an `ExchangeContext` to send/receive messages.

kghost commented on examples/chip-tool/commands/common/EchoCommand.cpp:
```diff
@@ -78,20 +80,35 @@ void EchoCommand::ReceiveEcho(PacketBuffer * buffer) const
+    em = dc->GetExchangeManager();
+    VerifyOrExit(em != nullptr, err = CHIP_ERROR_NO_MEMORY);
+
+    ec = em->NewContext(remoteId, this);
+    VerifyOrExit(ec != nullptr, err = CHIP_ERROR_NO_MEMORY);
```
Client (AKA controller) needs to manually create an `ExchangeContext` to send/receive messages. (EchoCommand)

bzbarsky-apple replied:
Just to be clear, "EchoCommand" is not the normal case we should be looking for here....

Better to think about the "on" command.  Let's say a switch is flipped and wants to send the "on" command to "the things that are supposed to be switched by this switch".  How does it do that?  That is what bindings address.

kghost commented on examples/chip-tool/commands/common/ModelCommand.cpp:
```diff
@@ -46,11 +46,22 @@ bool isGlobalCommand(uint8_t frameControl)
 
 void ModelCommand::OnConnect(ChipDeviceController * dc)
 {
+    mExchangeContext = dc->GetExchangeManager()->NewContext(mRemoteId, this);
```
Client (AKA controller) needs to manually create an ExchangeContext to send/receive messages. (ModelCommand)

bzbarsky-apple replied:
Yes, before initiating a data model transaction (which maps to a message layer exchange), you have to create an ExchangeContext.  This part we agree on.

kghost commented on examples/common/chip-app-server/Server.cpp:
```diff
@@ -45,26 +45,21 @@ using namespace ::chip::DeviceLayer;
 public:
-    void OnMessageReceived(const PacketHeader & header, const PayloadHeader & payloadHeader, Transport::PeerConnectionState * state,
-                           System::PacketBuffer * buffer, SecureSessionMgrBase * mgr) override
+    void OnMessageReceived(ExchangeContext * ec, const PacketHeader & packetHeader, uint32_t protocolId, uint8_t msgType, System::PacketBuffer * buffer) override
```
server (AKA device) will receive the allocated `ExchangeContext` here. The context is allocated by `ExchangeManager` when receiving a message with `Initiator` flags set.

bzbarsky-apple replied:
Yes, this is all fine so far.

But the low-level send functions need to deal with transactions initiated by the server too.

kghost commented on examples/common/chip-app-server/Server.cpp:
```diff
@@ -125,7 +121,9 @@ void InitServer()
-    gSessions.SetDelegate(&gCallbacks);
+    gExchangeManager.Init(&gSessions);
+    gSessions.SetDelegate(&gExchangeManager);
+    gExchangeManager.RegisterUnsolicitedMessageHandler(Protocols::kProtocol_InteractionModel, 0, &gCallbacks);
```
Server register an `ExchangeContextDelegate` to handle callbacks.

kghost reviewed (commented):
Highlight essential flows

kghost commented:
Here is another issue we may want to resolve. https://github.com/project-chip/connectedhomeip/issues/3642

kghost commented:
@bzbarsky-apple I want to move this forward, I know the point that the proposed method to create an exchange context may not be the best choice, but using node id to send/receive message is absolutely not a good idea. I want to move forward step by step. First step, as I proposed here, make changes that applications (ember is an application) use `ExchangeContext` to send and receive messages.

The concept of binding is to initiate a secure connection and establish an `ExchangeContext`, and currently the concept is not implemented at all, we can discuss it later.

kghost commented:
@vivien-apple Add Vivien to the discussion

bzbarsky-apple commented:
I think the right way to move forward step by step is to do what I suggested above, instead of removing code that we're going to have to add back and writing code that we're definitely going to need to rip out.

If you want a stopgap for now to just enable `ExchangeContext` in general and not be blocked on the data model code, one option that I would consider more palatable than what we have here is to just change the places that currently stash the `SecureSessionMgrBase` and implement `SessionManager()` to stash the `ExchangeContext` and have a hacky function to return that, instead of the hacky function to return the session manager.  Then `chipSendUnicast` will just ignore the `NodeId` argument, and we will need to work on fixing that, but in the meantime you will be unblocked.

kghost commented:
Please take a look at what I proposed, I think this a the direction we are going. I didn't remove any code which is going to be added back. I suggest that all applications above exchange layer uses `ExchangeContext` to send and receive messages.

All codes I have removed is dead features or not implemented features, I removed then because they are not working at all, so I have no way to fix them.

What you suggest is adding lots is code which is going to be removed later, and we may want to avoid.

bzbarsky-apple commented:
I did read what you proposed, of course; this whole conversation started with me reading the changeset.

Again, in case it was not clear: The changes to `af-main-common.cpp` are removing code that we absolutely want to be using in the future, including code to originate a multicast message to a specific group id or a binding and code to originate a unicast message to a binding.  These are both situations in which there is no existing `ExchangeContext` and the code in question will need to be modified to create one, but a lot of the other structure of the code will remain.

> I suggest that all applications above exchange layer uses ExchangeContext to send and receive messages.

Sure, but someone needs to create that `ExchangeContext` at some point for messages that are not responses to an incoming message, and this PR is removing the code in `af-main-common.cpp` that would be doing that creation!

> All codes I have removed is dead features or not implemented features

They are not yet completely working in the CHIP setup, but people are working on fixing that, in general.  And we absolutely need these features to ship CHIP.  Again, do not remove this code.

> so I have no way to fix them.

I am not asking you to fix them.  I am asking you to not make changes that are incompatible with the ongoing work of fixing them.

> What you suggest is adding lots is code which is going to be removed later

We already have a bunch of code that needs to be removed sometime, possibly as part of this PR: the various `SessionManager()` functions.   Those are a stopgap that needs to go away, in general.  It was put in place because we needed to be testing cluster code and the core secure channel APIs were not ready in their finalized form at that point.

Let me try again to describe what I see as reasonable options here:

1. What I see as a non-stopgap solution: go through the various places where we currently pass a pair of arguments like `EmberOutgoingMessageType type, uint64_t indexOrDestination` and convert them to something where we pass in a single type that identifies a destination.  Under the hood that single type will presumably have a union (of `ExchangeContext*`, binding index, group id seems like a plausible set of things) and an enum that says what to do with the union.  Modify the code of all the relevant functions to do the right thing with the passed-in value (pass it along, look up bindings, create `ExchangeContext`s, etc).  Pass through `ExchangeContext*` as needed.  Remove the current `SessionManager()` global function hack.  This is a fair amount of work, and I understand why you might not want to do it; I'm not asking you to, but this is the general direction this code will need to end up in eventually.
2. Semi-stopgap solution: keep using the two separate arguments and `uint64_t` as the effective union type, as now.  `static_assert` that `uint64_t` can hold all `uintptr_t` values.  Replace entrypoints into the sending code that currently take a node id with ones taking  a `uint64_t` or an `ExchangeContext*`, depending on what's more convenient.  At the point when we need to go from `ExhangeContext*` to `uint64_t`, do that by casting to `uintptr_t`.  In the code that handles the `EMBER_OUTGOING_DIRECT` case, and where right now we always treat the incoming `uint64_t` as a `NodeId`, cast it it to `ExchangeContext*` instead (possibly via `uintptr_t` for maximum clarity).  In the `EMBER_OUTGOING_VIA_BINDING` case, either set up an `ExchangeContext` for the binding's node id and use it to send, or comment out the `chipSendUnicast` call for now (since it will presumably take an `ExchangeContext*`) and file a followup issue to set up an `ExchangeContext`; others can take it from there.  Remove the current `SessionManager()` global function hack.  This will need an amount of changes in the data model code approximately on the order of what the PR has right now: the various places that became `void*` will just become `uint64_t` or `ExchangeContext*` for the moment.  I don't see any reason at all to use `void*` when we really mean `ExchangeContext*`, by the way.  It's possible that there was still some C, not C++, code involved when this PR was created, but at this point everything should be C++.
3. Quick-fix very stopgap solution: what I described in <https://github.com/project-chip/connectedhomeip/pull/3602#issuecomment-727135126>.  Change none of the data model APIs, replace `SessionManager()` with something to get the `ExchangeContext`, file an issue to implement option 1 above ASAP, because the setup will be really quite broken in various ways.

I'm really OK with you picking any of these options, or proposing other options that generally align with the expected code evolution described by option 1.  What I'm not really OK with is changes that actively make option 1 harder to implement in the future.

I'm happy to try to talk this through in some sort of synchronous setting if you would prefer that to going back and forth in PR comments.  Please let me know!

kghost commented:
Thank for your help. It is really helpful.

I'd like to pick option 1, with a little change.

> Under the hood that single type will presumably have a union (of ExchangeContext*, binding index, group id seems like a plausible set of things) and an enum that says what to do with the union.

I want to make the `ExchangeContext` to be the union. Move `binding index, group id seems like a plausible set of things) and an enum that says what to do with` into `ExchangeContext` as fields like this:

```
class ExchangeContext {
  ...
  enum BindingType { ... } mBindingType;
  int bindingIndex
  int groupId;
  ...
}
```

Then we can use only pointer to the `ExchangeContext`, to replace `indexOrDestination` as I proposed in my patches.

Would you please take consider of this approach. Thanks for your opinion.

bzbarsky-apple commented:
Hmm.  That is a good question.  I just looked over the various code that does `indexOrDestination` bits and it looks like it's either just passing it through or is the `send` method (or `emberAfSend`, which is only called from `send).  So it would not be unreasonable to say that the callers of `send` should allocate an `ExchangeContext` and pass it into `send`, instead of having `emberAfSend` do the allocation.  For now, I'm not even sure we'd need to add that extra state to it; as long as we can pass the `ExchangeContext` around and use it to send the actual message, that should be enough.

We will need a bit of thought for how to handle the refcounting/lifetime issues, and we will need a way to allocate a groupcast `ExchangeContext` at some point (and comment out the relevant `send` call for now, I guess); I am not seeing such a way right now on the `ExchangeManager`.   But yes, this might well be better than passing the union thing around, in that it will make it clearer which things are really part of the same exchange as opposed to coincidentally sharing the same destination.

kghost commented:
Great, I'll try this way first,

kghost commented:
@bzbarsky-apple Would you please check this

https://github.com/kghost/connectedhomeip/blob/ember/src/app/util/context.h

I'm going to add 2 classes `EmberContext` and `EmberSession`

`EmberSession` is a long live object associated with `ExchangeContext` to store session like members.

`EmberContext` is a short live object allocated on stack to eliminate all global variable currently used in ember library.

bzbarsky-apple commented:
I will look at the code later today, but:

1. We don't want to use "Ember" in new type names.  We are trying to get rid of it from old type names, in fact.
2. There is no "ember library".  There's just the CHIP library that code contributed by Silicon Labs is compiled into.

kghost commented:
> 1. We don't want to use "Ember" in new type names.  We are trying to get rid of it from old type names, in fact.
> 2. There is no "ember library".  There's just the CHIP library that code contributed by Silicon Labs is compiled into.

Understand, but we need a name to refer this part of codes, can we call it `InterchangeModel` or IM for short.

bzbarsky-apple commented:
Other things have been calling it DataModel, I think.

bzbarsky-apple commented:
OK, so looking at the code in `context.h`, some thoughts:

1. It looks like `EmberSession` more or less corresponds to an interaction model Transaction (which are 1-1 with exchanges), right?  If so, calling it `IMTransaction` or `DataModelTransaction` might make sense.  And there should probably be some documentation for what `mBinding` means or should be on the acceptor side...
2. For `EmberContext`, I think the reason the current code uses globals is to minimize the amount of dynamic allocation, including stack-allocation, and ensure that we have tight control over the memory usage of the application.  Is there a specific problem with the globals, other than "globals are not nice"?  There is no multi-threading involved for the message sending and receiving (at least right now), right?  I understand that we want to pass through the `ExchangeContext` and the session/transaction information, but it seems that we could just as easily add globals for those for the moment and then think about how, and whether, to change the globals setup around.  That would keep the scope of this PR narrower, which seems like a good idea.

kghost commented:
> OK, so looking at the code in `context.h`, some thoughts:
> 
> 1. It looks like `EmberSession` more or less corresponds to an interaction model Transaction (which are 1-1 with exchanges), right?  If so, calling it `IMTransaction` or `DataModelTransaction` might make sense.  And there should probably be some documentation for what `mBinding` means or should be on the acceptor side...

Good, I can rename it and move forward

> 2. For `EmberContext`, I think the reason the current code uses globals is to minimize the amount of dynamic allocation, including stack-allocation, and ensure that we have tight control over the memory usage of the application.  Is there a specific problem with the globals, other than "globals are not nice"?  There is no multi-threading involved for the message sending and receiving (at least right now), right?  I understand that we want to pass through the `ExchangeContext` and the session/transaction information, but it seems that we could just as easily add globals for those for the moment and then think about how, and whether, to change the globals setup around.  That would keep the scope of this PR narrower, which seems like a good idea.

1. The global variable doesn't reduce dynamic allocation, because the context is allocated on the stack, it is not dynamically allocated.

2. Some function also receive `indexOrDestination` as argument, can I remove it and just use the global variable when needed.

3. I agree that using globals is not causing any actual problem, only readability problem. And remove all globals beyonds scope of this PR. So I will leave it now.

kghost commented:
Another problem is `EmberOutgoingMessageType`. After being integrated to exchange, I think there will only one type of outgoing message, which is via `ExchangeContext`.

I will be forced to remove `EmberOutgoingMessageType`, because I can't integrate `EMBER_OUTGOING_VIA_BINDING` or `EMBER_OUTGOING_MULTICAST` into exchanges.

bzbarsky-apple commented:
> because the context is allocated on the stack, it is not dynamically allocated.

The stack is still dynamic allocation in that you don't know ahead of time how much memory (including stack) your app will use, and you can in fact run out of stack.

The benefit of the current setup is that you know statically, at compile time, how much memory at most will be needed for the data structures involved, and can ensure that fits on your hardware.

My understanding is that just like we are trying to eliminate heap-allocation in CHIP as much as we can, we are also trying to eliminate large stack allocations.

> Some function also receive indexOrDestination as argument, can I remove it and just use the global variable when needed.

In theory you could, I guess, but you'd then need to correctly set that global variable in various places.  The tradeoff doesn't seem worthwhile...

> And remove all globals beyonds scope of this PR

Indeed.

> After being integrated to exchange, I think there will only one type of outgoing message, which is via DataModelTransaction

Well, yes, the handling of the message type will presumably shift up to above where we get the appropriate transaction, right?

> I will be forced to remove EmberOutgoingMessageType,

This does not seem entirely implausible, if the getting of the right transaction/exchange happens higher up the callstack than the current uses of this argument.

> because I can't integrate EMBER_OUTGOING_VIA_BINDING or EMBER_OUTGOING_MULTICAST into exchanges.

The current consumers of `EMBER_OUTGOING_VIA_BINDING` should work just fine with exchanges, I'd think.  For example, `emberAfSendUnicastToBindingsWithCallback` should find the relevant binding, get the destination node id, start a new exchange (create a new `ExchangeContext`, etc), and then pass that down to the lower-level code in `send`.  Of course if the binding is not a unicast binding we can't do anything with that yet, just like now.

It looks like no one ever passes anything other than `EMBER_OUTGOING_DIRECT` to `emberAfSendUnicastWithCallback`, so that can presumably be changed to just take an `ExchangeContext` or `DataModelTransaction` or whatever to start with.

For `EMBER_OUTGOING_MULTICAST`, we're going to need to add support for creating an exchange for a groupcast, of course, which is not implemented yet.  Just like the current code fails out in that case, you can make the relevant entrypoints fail out for now.  The unfortunate part is that right now all the so-far-unimplemented parts are in one place (`emAfSend`) and in the new setup they will be scattered about more; it's worth thinking a bit about how to mitigate that, via TODO comments pointing to issues that get filed, etc.

kghost commented:
> In theory you could, I guess, but you'd then need to correctly set that global variable in various places. The tradeoff doesn't seem worthwhile...

There are only 2 places I need to set that global variable
1. When receiving a new message, I need to adjust the global `ExchangeContext` pointer to the corresponding exchange.
2. When a timer triggered, for example, a reporting timer. I need to set the context to send the report.

They are both at the beginning of handling something, so setting the global is reasonable here. I think it is better than passing them all way along the call stack.

bzbarsky-apple commented:
> There are only 2 places I need to set that global variable

You're missing at least one more: when we get `emAfMessageSentHandler`.  Which is not called from anywhere yet, but that still needs to be fixed in the CHIP stack.  And of course your second place is, in practice, a number of different places, which will need pretty similar code in the end.

I'm willing to be proved wrong by demonstration here, by the way, but I suspect there will be a lot more whack-a-mole going on here, and less clarity about whether we're using the right `ExchangeContext` and how we tell whether we are than if it gets passed through explicitly (possibly as part of another struct).

To the larger point about globals: I agree that ideally we would not have random globals being poked from various places because it makes it hard to tell who is using what resources when.  But I'd rather carefully design how to get rid of that while keeping the major memory usage completely static, in a separate PR that lets us measure the codesize and memory impact and evaluate the change.  Doing it all as part of this PR doesn't really make that possible, so I would like us to not change the existing global accesses any more than we have to as part of making this change.  If you want to reuse the current "response node id" global for `ExchangeContext` or the like, that's probably OK for the moment.  But again, I'd have to look at the resulting code to see how it looks, and we'd need to add some sets (and unsets) of that global in various places.

kghost commented:
I will use `DataModelContext` to replace `indexOrDestination` and leave the global problem later.

bzbarsky-apple commented:
> I will use DataModelContext & to replace indexOrDestination and leave the global problem later.

Again, I'll have to see what the code looks like, but I'd think the right replacement for `indexOrDestination` would be something representing the transaction, right?  So naming-wise `DataModelTransaction` might be a better fit, but we'll see what the code looks like.

> I'll have also removed multicast and broadcast APIs, they will be implemented in ExchangeContext in the future. 

It makes the most sense to me to _keep_ those APIs for now (if you're talking about things like `emberAfSendMulticastWithCallback`), and in the future create the new ExchangeContext (because those APIs imply initiation of a transaction) inside those APIs.  For now, of course, they will just need to error out, with TODO comments and followup issues filed.  Again, maybe I am just misunderstanding the proposal; we'll see how the code comes out.  But I expect that controllers may want to initiate communications directly with a specific node id or group id, not via bindings, depending on exactly how we set up controllers.  So I would rather minimize the code removal until we get that sorted out.

bzbarsky-apple reviewed (commented):
There seem to be merge issues in various `gen` files, with this PR backing out changes from other PRs?

bzbarsky-apple commented on src/app/util/af-main-common.cpp:
```diff
@@ -269,7 +267,7 @@ static EmberStatus send(EmberOutgoingMessageType type, uint64_t indexOrDestinati
     {
         EmberAfMessageStruct messageStruct = {
-            callback, apsFrame, message, indexOrDestination, messageLength, type, broadcast,
+            callback, apsFrame, message, messageLength,
```
This should presumably get the ExchangeContext.

bzbarsky-apple commented on src/app/util/af-main-common.cpp:
```diff
@@ -284,9 +282,9 @@ static EmberStatus send(EmberOutgoingMessageType type, uint64_t indexOrDestinati
+    if (messageLength <= dmContext.GetExchangeContext().GetMTU())
     {
-        status = emAfSend(type, indexOrDestination, apsFrame, (uint8_t) messageLength, message, &messageTag, alias, sequence);
+        status = chipSendUnicast(exchangeContext, apsFrame, messageLength, message);
```
TODO comment to rename the API, linking to the issue tracking that, please.

bzbarsky-apple commented on src/app/util/af-main-common.cpp:
```diff
@@ -284,9 +282,9 @@ static EmberStatus send(EmberOutgoingMessageType type, uint64_t indexOrDestinati
     emAfApplyRetryOverride(&apsFrame->options);
 
-    if (messageLength <= emberAfMaximumApsPayloadLength(type, indexOrDestination, apsFrame))
+    if (messageLength <= dmContext.GetExchangeContext().GetMTU())
```
This is, unfortunately, not the right check.  I assume that `GetMTU` returns the size of the allowed CHIP application payload (though that's not what's implemented yet), but that payload will consist of the messageLength payload bytes here plus the framing that `chipSendUnicast` adds.  So we need to do something better here.  TODO with tracking issue is OK, and please tell me whether you plan to work on that  issue.

kghost replied:
Got it.

bzbarsky-apple commented on src/app/util/af-main-common.cpp:
```diff
@@ -335,129 +333,18 @@ static EmberStatus send(EmberOutgoingMessageType type, uint64_t indexOrDestinati
-                alias, sequence, callback);
-}
-
-EmberStatus emberAfSendMulticastWithCallback(GroupId multicastId, EmberApsFrame * apsFrame, uint16_t messageLength,
```
Again, please do not remove these APIs for now.

bzbarsky-apple commented on src/app/util/af-main-common.cpp:
```diff
@@ -335,129 +333,18 @@ static EmberStatus send(EmberOutgoingMessageType type, uint64_t indexOrDestinati
-    return emberAfSendMulticastWithCallback(multicastId, apsFrame, messageLength, message, NULL);
-}
-
-EmberStatus emberAfSendMulticastToBindings(EmberApsFrame * apsFrame, uint16_t messageLength, uint8_t * message)
```
And _definitely_ do not remove this API!  AS you might have noticed in `reporting.cpp`, it's needed for reporting as currently implemented; changing from "send to all the multicast bindings" to "send to all the unicast bindings" is quite a big semantic change: it sends to a different sent of destinations!

kghost replied:
The thing which confuse me it that, eventually `SendMessage` of all types of message will be handled by `ExchangeContext::SendMessage`. There will be unicast exchanges, which send unicast messages, there will be multicast exchanges, which send multicast messages, and it is totally transparent to DataModel. The type is exchanges can be fetched using `ExchangeContext::IsUnicast` or `ExchangeContext::IsMulticast` or `ExchangeContext::IsBroadcast`. The type of exchange is decided when creating the exchange.

So do we really need separate APIs for each type of message ? If we keep these APIs, they will both become `ExchangeContext::SendMessage`

kghost replied:
I'm fine to add them back. They are not causing any trouble if we just ignore them.

bzbarsky-apple replied:
I think we should probably sort out the questions in https://github.com/project-chip/connectedhomeip/pull/3602#discussion_r526985171 before we get too far into the discussion here, but the way I see it, the operations the data model needs from the message layer are:

1. Respond to this incoming message, using the exchange it came in on.
2. Create a new exchange for a given node id, do some message sending/receiving on it.
3. Create a new exchange for a given group id, do some message sending (and maybe receiving) on it.
4. Tear down an exchange, whether we created it or whether it was created by an incoming message, once we expect to not send/receive any more messages on it.

> and it is totally transparent to DataModel

For items 2 and 3, data model needs to _create_ the exchange when it wants to start a new transaction, right?  That is what the APIs involved should do: data model code should be able to say "hey, I want to start a new multicast transaction" and we should have some single function or small set of functions (depending on how the target of the multicast is being identified: bindings vs raw group id), which does the work involved in that: setting up the new exchange and then sending the message.

bzbarsky-apple replied:
> I'm fine to add them back. They are not causing any trouble if we just ignore them.

OK, good.  Let's leave them for now, then.  :)

bzbarsky-apple commented on src/app/util/context.h:
```diff
@@ -0,0 +1,138 @@
+    void OnExchangeClosing(chip::ExchangeContext * ec) override;
+
+private:
+    // TODO: is there any member belong to the transaction ?
```
Yes, I would expect the `ExchangeContext` here, no?  And then we don't need a separate `DataModelContext`....

The ownership (who frees what when) situation will, of course, need to get sorted.  It's really unclear so far for `ExchangeContext` how that's supposed to work...

kghost replied:
The `ExchangeContext` doesn't belong here. Because `ExchangeContext` will be always passed in as an augment.

If there is nothing in the `Transaction`, I can remove this class. `Transaction` should be dynamically allocated or use a object pool, it is better that we don't need it.

bzbarsky-apple replied:
> ExchangeContext will be always passed in as an augment.

When _receiving_ messages.  How do we represent sending outgoing exchange-initiating messages?

I'd basically like to have some representation of the concept of a "data model transaction".  Per spec, this happens to be synonymous with a "message layer exchange".  Is there an existing object that represents that thing?  My impression was that `ExchangeContext` was it, but if not, we need to figure out what that is.  If that _is_ it, then it can't quite represent a transaction, because it needs a delegate to handle receipt of messages (is that true?), so a transaction needs to encapsulate that delegate and the `ExchangeContext` somehow.

The things a "data model transaction" needs to support are:

1. Sending initial message when transaction is created.  This needs the `ExchangeContext` at first.
2. Receiving messages and responding to them.  This needs an `ExchangeDelegate`.
3. Controlling the lifetime of the exchange: closing it on "last message" send, possibly on timeouts, etc.  This is doable with just the delegate if the timeout bits are handled entirely in the message layer.  If the timeout logic lives in the interaction model layer, then it can't just use the delegate, afaict.

bzbarsky-apple commented on src/app/util/types_stub.h:
```diff
@@ -538,16 +502,13 @@ typedef struct
-        chip::GroupId groupId;
-    };
+
+    /** The exchange associated with the binding. Messages are send or received
```
No.  No.  Bindings have node or group ids.  The CANNOT have a transaction or exchange associated with them.  Bindings are long-lived data structures that survive reboot and the like.  Transactions/exchanges are short-lived.  Furthermore, there might be multiple transactions at the same time that are all based on the same binding!

kghost replied:
In this case we need to translate these persistent bindings into active binding, where `ExchangeContext` are stored. We can't send message only given node ids.

bzbarsky-apple replied:
OK, so a few things.  Maybe I completely misunderstand the scope of an `ExchangeContext`?  My understanding was that there was one `ExchangeContext` per exchange, with it being created when the exchange starts and destroyed when the exchange ends.  Is that correct?

If that is correct, then the right flow would seem to be that when the data model wants to start a new transaction (which corresponds to an exchange) it needs to _at that point_ create an `ExchangeContext` and send via that context.

Is my understanding of the interaction between exchanges and `ExchangeContext` incorrect?  If not, is my understanding that a new exchange is cheap and quick to set up (which a bunch of the existing specification text is based on) incorrect?  If we have to do mDNS discovery for every single exchange setup, that doesn't seem obviously acceptable for the way the current interaction model spec proposes to use exchanges.

My understanding from the last time I spoke to folks about the APIs data model should deal with in terms of interacting with lower layers was that it would go about as follows:

1. For incoming messages and their replies, there would be a representation of the exchange that came with the message that could be used to send a reply.  Pretty straightforward.
2. For message origination, the data model would hand a node id or group id to the lower layers and get back a representation of a newly created exchange.  It would then be responsible for managing the lifetime of that exchange and send messages pertaining to one data model transaction through it.  When the transaction is done, the exchange would be released.  Any details about secure channel establishment, determining and caching mappings from node ids to ips, etc, would be an implementation detail of the message layer.

Has the API contract changed since then?  If so, can you please describe the new API contract?

> We can't send message only given node ids.

For an outgoing message, given the current spec setup, we _must_ on the data model level, be able to do precisely that, by creating a new exchange from that node id (which might require establishment of a secure channel if one is not already established, of course, but that's a message layer implementation detail) and then sending messages via the exchange.

kghost replied:
Yes, you are absolutely right. exchanges are short-live and cheap. But lots of prerequisite is needed. If it doesn't require an exchange, then it still need a structure like `Cached<PeerAddress, KeyId, SecureSessionState>` to send messages.

If we leave the caching mechanism to underlayer implementations totally, it will probably fail to send several reports when something happens, like device reboots, peer IP address changes, session key rotates. With the `ExchangeContext`, the data model is able to acknowledge `OnConnectionClosed` event, and re-establish the channel as soon as possible.

bzbarsky-apple replied:
Maybe that's the heart of the disagreement about how this should work?  My understanding was that channels are expensive to track and a node can potentially need way more channels than it can support in-flight at once, so channels are set up on an as-needed basis, not constantly maintained.  We definitely need to sort this out, but it seems to be out of the scope of this PR as well...  

In any case, it seems to me that what ends up happening when we want to do a new exchange is that we should be able to ask the underlying layer for a new exchange.  Sometimes that will require channel establishment, because we can't keep all channels open all the time, in my understanding, so whatever APIs and data flows we design around this need to be able to handle that.  If there is channel caching and keeping alive going on, it should be separate from the basic mechanism of "ok, we want to send to this node id; look up the channel for it or create one if there isn't one yet"....

If I am totally wrong and the intent _is_ in fact to keep all possible outgoing channels active at all times, then obviously that changes our options for the design of the sending APIs in data model.

kghost commented:
> There seem to be merge issues in various `gen` files, with this PR backing out changes from other PRs?

I'm using templates in master branch, and zap file from this PR https://github.com/project-chip/connectedhomeip/pull/3610

bzbarsky-apple commented:
Why are the `gen` files being touched at all?

If they have to be touched, please talk to @vivien-apple about the right way to do it.

kghost commented:
> Why are the `gen` files being touched at all?
> 
> If they have to be touched, please talk to @vivien-apple about the right way to do it.

They are using `indexOrDestination` argument, I need to change it to the context.

https://github.com/project-chip/connectedhomeip/pull/3602/commits/695ba796a9920e67a70683cc695f89d241498f82#diff-f54ad7ff3d9777ab36df9ddc155db613e156a172f9d0e0d275c4de11165ae0ce
https://github.com/project-chip/connectedhomeip/pull/3602/commits/695ba796a9920e67a70683cc695f89d241498f82#diff-aa798805d96ac1c862db94481e801f8fb689e1437c5cf099c268b14accf6a6c2

bzbarsky-apple commented:
> They are using indexOrDestination argument, I need to change it to the context.

Ah, I see.  So the key thing is: the `gen` files are NOT yet being generated from the `.zapt` files, as I understand it.  So at the moment you should probably change the `gen` files manually if they need changing.  The changes to the `.zapt` files can either happen here too, or in a followup.

kghost commented:
I think we have resolved most of the problems. The last problem is the binding table. The active binding must contains an `ExchangeContext`, but the persistent bindings contains node id.

I think we can store persistent bindings in flash, and keep active bindings in memory. The current structure `EmberBindingTableEntry` is representing active bindings in memory. We need another API to create these bindings, probably provided by `ExchangeManager`

bzbarsky-apple commented:
> The active binding must contains an ExchangeContext

Given that a "binding" in the data model sense can correspond to multiple different in-flight exchanges at once, it sounds like you and I have very different understandings of the lifetime of an `ExchangeContext` and how it corresponds to "an exchange".  I'm assuming my understanding is wrong, as are the comments in `ExchangeContext.h`, and would really appreciate an explanation of how these things relate to each other.

kghost commented:
> Given that a "binding" in the data model sense can correspond to multiple different in-flight exchanges at once, it sounds like you and I have very different understandings of the lifetime of an `ExchangeContext` and how it corresponds to "an exchange". I'm assuming my understanding is wrong, as are the comments in `ExchangeContext.h`, and would really appreciate an explanation of how these things relate to each other.

I think you are right. Currently I have no idea how to smoothly create an exchange and then send a message. It is not an easy task to turn a node id into an exchange,  but requires lots of initialization. We may need to give a hint to the `ExchangeManager` that it should always prepare a channel to a group or a node for sending message. And we use the handle to to channel to replace `ExchangeContext` in the binding table.

bzbarsky-apple commented:
> It is not an easy task to turn a node id into an exchange, but requires lots of initialization

OK.  This seems like the main thing we need to solve at some point here....

> But even we implement it, it is probably not cheaper than the ExchangeContext. So I'm not sure if it is fine to just use an ExchangeContext here.

Where is "here" in this case?

kghost commented:
I think you are right, I agree that we use node id or group id in the binding table. And I'll try provide another class or API to "resolve" the node id or group id into a new exchange. Maybe I can find a solution later, until then, we can hold this PR.

bzbarsky-apple commented:
OK.  Again, if this is blocking other work, I am quite happy to make changes to accommodate that work happening, along some of the lines outlined in <https://github.com/project-chip/connectedhomeip/pull/3602#issuecomment-727519125>.  Please just let me know.

woody-apple commented:
@kghost @andy31415 What are the next steps here?

kghost commented:
@woody-apple 

I'll get #4019 merged first, then use the Channel API to implement missing parts of this PR.

kghost commented:
I'll create another PR when Channel is ready. Close this one.
